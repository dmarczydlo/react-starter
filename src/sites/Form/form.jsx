import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Form extends Component {

    state = {
        error: null,
        name: ""
    };

    static getDerivedStateFromProps(nextProps, prevState) {
        if (nextProps.form.error && nextProps.form.error !== prevState.error) {
            return { error: nextProps.form.error};
        }

        return null;
    }

    handleOnChange = event => {
        this.setState({ name: event.target.value });
    }

    handleOnSubmit = () => {
        const { name } = this.state;
        const { sendDataAPI } = this.props;

        if (!name) {
            this.setState({ error: 'Name can\'t be empty' });
        } else {
            sendDataAPI(name);
            this.setState({ error: '' });
        }
    }

    render() {
        const { error } = this.state;
        const { form: { data, sending } } = this.props;
        return (
            <fieldset>
                <h2>
                    {'Form'}
                </h2>
                {error && (
                    <p>
                        {error}
                    </p>)
                }
                <label htmlFor="name">
                    {'Name'}
                </label>
                <input
                    name="name"
                    onChange={this.handleOnChange}
                    type="text"
                />
                <button
                    id="button"
                    onClick={this.handleOnSubmit}
                    type="button"
                >
                    {'Send to API'}
                </button>
                {data && !sending && (
                    <p>
                        {JSON.stringify(data)}
                    </p>)
                }
            </fieldset>
        );
    }
}

Form.propTypes = {
    form: PropTypes.shape({
        data: PropTypes.object.isRequired,
        sending: PropTypes.bool.isRequired
    }).isRequired,
    sendDataAPI: PropTypes.func.isRequired
};
export default Form;
