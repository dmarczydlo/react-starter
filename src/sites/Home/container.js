import Site from './home';
import { connect } from 'react-redux';
import { updateValue } from '../../actions/counter';

export const mapStateToProps = (state) => {
    return state;
};

export const mapDispatchToProps = (dispatch) => {
    return { updateValue: type => dispatch(updateValue(type)) };
};

export default connect(mapStateToProps, mapDispatchToProps)(Site);
