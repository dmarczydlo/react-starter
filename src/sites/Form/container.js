import Site from './form';
import { connect } from 'react-redux';
import { sendDataAPI } from '../../actions/form';

export const mapStateToProps = (state) => {
    return state;
};

export const mapDispatchToProps = (dispatch) => {
    return { sendDataAPI: name => dispatch(sendDataAPI(name)) };
};

export default connect(mapStateToProps, mapDispatchToProps)(Site);
