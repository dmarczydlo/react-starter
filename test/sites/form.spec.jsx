import React from 'react';
import sinon from 'sinon';
import FormContainer from '../../src/sites/Form';
import Form from '../../src/sites/Form/form';
import { mapDispatchToProps, mapStateToProps } from '../../src/sites/Form/container';

describe('Form site', () => {
    let wrapper = null;
    const form = { data: {} };
    const mockDispatchFunc = sinon.spy();

    const sendDataAPI = jest.fn();

    const store = {
        data: {},
        error: '',
        sending: false
    };

    it('form container should be rendered', () => {
        wrapper = shallow(<FormContainer />);
        expect(wrapper).toMatchSnapshot();
    });

    it('form  should be rendered', () => {

        wrapper = shallow(
            <Form
                form={store}
                sendDataAPI={sendDataAPI}
            />);
        expect(wrapper).toMatchSnapshot();
    });


    it('button should can be clicked', () => {
        const mockCallBack = sinon.spy();
        wrapper = mount((
            <Form
                form={store}
                sendDataAPI={mockCallBack}
            />));
        wrapper.setState({ name: 'bar' });
        // eslint-disable-next-line no-magic-numbers
        wrapper.find('#button').at(0).
            simulate('click');
        // eslint-disable-next-line no-magic-numbers
        expect(mockCallBack).toHaveProperty('callCount', 1);
    });


    it('error should be set', () => {
        const mockCallBack = sinon.spy();
        wrapper = mount((
            <Form
                form={store}
                sendDataAPI={mockCallBack}
            />));
        // eslint-disable-next-line no-magic-numbers
        wrapper.find('#button').at(0).
            simulate('click');
        // eslint-disable-next-line no-magic-numbers
        expect(wrapper.state().error).not.toBeNull();
        // eslint-disable-next-line no-magic-numbers
        expect(mockCallBack).toHaveProperty('callCount', 0);
    });

    it('should call name onChange prop', () => {
        const mockCallBack = sinon.spy();
        const name = "name";
        const event = {
            target: {
                name,
                value: name
            }
        };
        wrapper = mount(
            <Form
                form={store}
                sendDataAPI={mockCallBack}
            />);


        wrapper.find('input').simulate('change', event);
        expect(wrapper.state().name).toEqual(name);
    });

    it('getDerivedStateFromProps ', () => {
        const error = "New error";
        const givenProps = { form: { error } };
        const givenState = { error: '' };

        const result = Form.getDerivedStateFromProps(givenProps, givenState);
        expect(result).toEqual({ error });
    });

    it('mapStateToProps should can be used', () => {
        // eslint-disable-next-line no-magic-numbers
        expect(mapStateToProps(form).data).toEqual({});
    });

    it('mapDispatchToProps should can be used', () => {
        mapDispatchToProps(mockDispatchFunc).sendDataAPI('abcd');
        // eslint-disable-next-line no-magic-numbers
        expect(mockDispatchFunc).toHaveProperty('callCount', 1);
    });
});
