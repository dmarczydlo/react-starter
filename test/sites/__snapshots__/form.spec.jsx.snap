// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form site form  should be rendered 1`] = `
<fieldset>
  <h2>
    Form
  </h2>
  <label
    htmlFor="name"
  >
    Name
  </label>
  <input
    name="name"
    onChange={[Function]}
    type="text"
  />
  <button
    id="button"
    onClick={[Function]}
    type="button"
  >
    Send to API
  </button>
  <p>
    {}
  </p>
</fieldset>
`;

exports[`Form site form container should be rendered 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
